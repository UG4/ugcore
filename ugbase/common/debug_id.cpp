/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#include "debug_id.h"
#include "common/log.h"
#include "common/error.h"
#include "common/assert.h"
#include <iostream>
#include <string.h>
#include "util/string_util.h"

namespace ug{

/**
 * register the debug id.
 * NOTE: this function is called in the initialization of variables
 * of type DebugID, which are mostly global variables.
 * Be absolutely sure we are safe here, i.e.
 * we are not using other things which might not be initialized (like Log).
 */
DebugID::DebugID(const char *str)
{
	m_hash = crc32(str);
	GetDebugIDManager().register_debug_id(str);
}

DebugIDManager& DebugIDManager::instance()
{
	static DebugIDManager m;
	return m;
}


bool DebugIDManager::
set_debug_levels(int lev)
{
	for(std::map<uint32, int>::iterator it=m_dbgLevels.begin(); it != m_dbgLevels.end(); ++it)
		(*it).second = lev;
	return true;
}

bool DebugIDManager::
set_debug_level(const char *debugID, int level)
{
	int slen = strlen(debugID);
	if(slen<=0) return false;
	if(debugID[slen-1] == '*')
	{
		for(size_t i=0; i<m_dbgLevelIdentifiers.size(); i++)
		{
			const char *name = m_dbgLevelIdentifiers[i].c_str();
			if(WildcardMatch(name, debugID))
			{
				set_debug_level(crc32(name), level);
				//UG_LOGN(name);
			}
		}
	}
	else if(set_debug_level(crc32(debugID), level) == false)
	{
		UG_LOG("DebugID " << debugID << " not registered.\n");
		return false;
	}
	return true;
}

/**
 * register the debug id.
 * NOTE: this function is called in the initialization of global variables
 * of type DebugID. Be absolutely sure we are safe here, i.e.
 * we are not using other things which might not be initialized (like Log).
 */
bool DebugIDManager::
register_debug_id(const char *debugID)
{
	if(debug_id_registered(debugID) == false)
	{
		m_dbgLevelIdentifiers.push_back(std::string(debugID));
		m_dbgLevels[crc32(debugID)] = -1;
		return true;
	}
	else
	{
		// not quite clear if cout is defined yet.
		// --> it should be, if we define it in this header!
		std::cout << "FATAL ERROR: DebugID "<<debugID<<" already registered." << std::endl;
		// note that this could be caused by double-registering libraries.
		UG_THROW("FATAL ERROR: DebugID "<<debugID<<" already registered.");
		return false;
	}
}

}
