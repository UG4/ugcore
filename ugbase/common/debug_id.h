/*
 * Copyright (c) 2012-2015:  G-CSC, Goethe University Frankfurt
 * Author: Martin Rupp
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#ifndef __H__UG__COMMON__DEBUG_ID_H
#define __H__UG__COMMON__DEBUG_ID_H
#include <vector>
#include <map>
#include <string>
#include "common/util/crc32.h"
#include <sstream>

namespace ug{

/// \addtogroup ugbase_common
/// \{

class DebugIDManager;



// DebugID
/**
 * A DebugID is an object to control the debug level of parts of the code.
 * It can be used in
 * - UG_SET_DEBUG_LEVEL
 * - UG_DEBUG_BEGIN/END (debugID, level)
 * - IF_DEBUG
 * - UG_DLOG(debugID, level, msg)
 *
 * ug4's standard DebugIDs are
 * APP, LIB_GRID, LIB_GRID_REFINER, LIB_DISC, LIB_DISC_ASSEMBLE, LIB_DISC_D3F,
 *		LIB_DISC_MULTIGRID, LIB_DISC_NEWTON, LIB_DISC_LINKER, LIB_DISC_TRANSFER,
 *		LIB_DISC_DISCRETE_FUNCTION, LIB_DISC_OUTPUT, LIB_DISC_OPERATOR_INVERSE,
 *		LIB_ALG_LINEAR_OPERATOR, LIB_ALG_LINEAR_SOLVER, LIB_ALG_VECTOR,
 *		LIB_ALG_MATRIX, LIB_ALG_AMG, LIB_PCL
 *
 * They are managed with \sa DebugIDManager.
 * You can define your own DebugID the following way:
 * In a cpp file:
 *
 \code{.cpp}
 DebugID DID_MYAPP("MyApp")
 DebugID DID_MYAPP_PARALLELIZATION("MyApp.Parallelization")
 \endcode{.cpp}
 *
 * Now to use it in other files you need to add there
 *
 \code{.cpp}
 extern DebugID DID_MYAPP, DID_MYAPP_PARALLELIZATION;
 ...
 UG_DLOG(DID_MYAPP, 3, "bla");
 \endcode{.cpp}
 *
 * Note that the DebugID OBJECT is called DID_MYAPP, but its string identifier
 * is "MyApp" (this is used in the script system). In LUA script you can do
 *
 * GetLogAssistant():set_debug_level("MyApp", 3)
 * or, with completion (UGIDE)
 * SetDebugLevel(debugID.MyApp, 3)
 *
 * \note The actual DebugID level is stored in DebugIDManager.
 *
 */
class DebugID
{
public:
/// registers the DebugID at DebugIDManager
	DebugID(const char *str);
	
	
	DebugID(uint32 hash) { m_hash = hash; }
/// returns the debug level via GetDebugIDManager.
	inline int get_debug_level() const;

/// sets the debug level via GetDebugIDManager.
	inline bool set_debug_level(int level);

	friend class DebugIDManager;
private:
	uint32 m_hash;
};

// the list of ug4's standard DebugIDs.
extern DebugID MAIN,
		APP,
		LIB_GRID,
		LIB_GRID_REFINER,
		LIB_DISC,
		LIB_DISC_ASSEMBLE,
		LIB_DISC_D3F,
		LIB_DISC_MULTIGRID,
		LIB_DISC_NEWTON,
		LIB_DISC_LINKER,
		LIB_DISC_TRANSFER,
		LIB_DISC_DISCRETE_FUNCTION,
		LIB_DISC_OUTPUT,
		LIB_DISC_OPERATOR_INVERSE,
		LIB_ALG_LINEAR_OPERATOR,
		LIB_ALG_LINEAR_SOLVER,
		LIB_ALG_VECTOR,
		LIB_ALG_MATRIX,
		LIB_ALG_AMG,
		LIB_PCL;


/**
 * The DebugIDManager.
 * get, set debug level, get a list of available DebugIDs etc.
 */
class DebugIDManager
{
private:
	/// returns the debug level of debugIDhash, -1 if not found
		inline int get_debug_level(uint32 debugIDhash) const
		{
			std::map<uint32, int>::const_iterator it = m_dbgLevels.find(debugIDhash);
			if(it == m_dbgLevels.end()) return -1;
			else return (*it).second;
		}
		
		/// sets the debug level of debugIDhash if registered to 'level'
		inline bool set_debug_level(uint32 debugIDhash, int level)
		{
			std::map<uint32, int>::iterator it = m_dbgLevels.find(debugIDhash);
			if(it == m_dbgLevels.end()) return false;
			else
			{
				(*it).second = level;
				return true;
			}
		}
public:
	///	returns a reference to the single instance of LogAssistant
		static DebugIDManager& instance();

	/// returns true if there is a DebugID with name debugIDname
		inline bool debug_id_registered(const char *debugIDname) const
		{
			return m_dbgLevels.find(crc32(debugIDname)) != m_dbgLevels.end();
		}
	/// returns the debug level of debugID, -1 if not found
		inline int get_debug_level(const DebugID &debugID) const
		{
			return get_debug_level(debugID.m_hash);
		}
	/// returns the debug level of debugID, -1 if not found
		inline int get_debug_level(const char *debugID) const
		{
			return get_debug_level(crc32(debugID));
		}
		
	/// 
		inline DebugID get_debug_id(const char *debugID) const
		{
			return DebugID(crc32(debugID));
		}
		
		inline DebugID get_or_create_debug_id(const char *debugID)
		{
			if(debug_id_registered(debugID)==false)
				register_debug_id(debugID);
			return DebugID(crc32(debugID));
		}
				

	/// sets the debug level of debugID if registered to 'level'
		inline bool set_debug_level(DebugID &debugID, int level)
		{
			return set_debug_level(debugID.m_hash, level);
		}
	/// sets the debug level of debugID if registered to 'level'
		bool set_debug_level(const char *debugID, int level);

	
	/// sets the debug level of all registered debugIDs to 'lev'
		bool set_debug_levels(int lev);

	/// returns a string describing all registered debugIDs.
		std::string get_registered_debug_IDs() const
		{
			std::stringstream str;
			str << "DebugIDs:\n";
			for(size_t i=0; i<m_dbgLevelIdentifiers.size(); i++)
				str << m_dbgLevelIdentifiers[i] << " : Level " << get_debug_level(m_dbgLevelIdentifiers[i].c_str()) << "\n";
			return str.str();
		}

		const std::vector<std::string> &get_registered_debug_IDs_arr() const
		{
			return m_dbgLevelIdentifiers;
		}



	bool register_debug_id(const char *debugID);

private:
	std::vector<std::string> m_dbgLevelIdentifiers;
	std::map<uint32, int> m_dbgLevels;
};

inline DebugIDManager& GetDebugIDManager()
{
	return DebugIDManager::instance();
}

inline int DebugID::get_debug_level() const
{
	return GetDebugIDManager().get_debug_level(*this);
}
inline bool DebugID::set_debug_level(int level)
{
	return GetDebugIDManager().set_debug_level(*this, level);
}

// TemporaryDebugLevel
/// a helper object to temporary set debug levels in a given scope
/**
 * A object of this class can be used to change a DebugID. The DebugID will
 * be re-set to the value before at destruction of the object.
 * This is especially useful if algorithms are calling other functions which
 * also have special DebugIDs levels. And you want to do a "debug debug".
 * example:
\code{.cpp}
  UG_SET_DEBUG_LEVEL(LIB_GRID, 1);

  // DebugID LIB_GRID is now at level 1
  {
  	  TemporaryDebugLevel tdl(LIB_GRID, 5);

  	  // DebugID LIB_GRID is now at level 5
  	  MyFunction();
  }

  // DebugID LIB_GRID is now again at 1.
\endcode
 * You can also specify differences:
 *
 \code{.cpp}
  TemporaryDebugLevel tdl(LIB_GRID, MY_DEBUG_ID, +1);
  // now LIB_GRID has the temporary debug level of MY_DEBUG_ID +1.
 */
#ifdef UG_ENABLE_DEBUG_LOGS
class TemporaryDebugLevel
{
	DebugID &did;

	int m_prevDebugLevel;
public:
	TemporaryDebugLevel(DebugID &debugID) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
	}

	/// change the debugID level in this scope
	TemporaryDebugLevel(DebugID &debugID, int temporaryDebugLevel) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
		did.set_debug_level(temporaryDebugLevel);
	}

	/// change the debugID level in this scope to another debugID plus some diff.
	TemporaryDebugLevel(DebugID &debugID, DebugID &debugSrc, int diff=0) : did(debugID)
	{
		m_prevDebugLevel = did.get_debug_level();
		did.set_debug_level(debugSrc.get_debug_level()+diff);
	}

	/// resets the debugID to the previous value.
	~TemporaryDebugLevel()
	{
		did.set_debug_level(m_prevDebugLevel);
	}

	void set(int temporaryDebugLevel)
	{
		did.set_debug_level(temporaryDebugLevel);
	}
};
#else
class TemporaryDebugLevel
{

public:
	TemporaryDebugLevel(DebugID &debugID){}
	/// change the debugID level in this scope
	TemporaryDebugLevel(DebugID &debugID, int temporaryDebugLevel){}
	/// change the debugID level in this scope to another debugID plus some diff.
	TemporaryDebugLevel(DebugID &debugID, DebugID &debugSrc, int diff=0){}
	/// resets the debugID to the previous value.
	~TemporaryDebugLevel(){}
	void set(int temporaryDebugLevel){}
};
#endif

// end group ugbase_common
/// \}

}	// end namespace ug
#endif	/* __H__UG__COMMON__DEBUG_ID_H */

