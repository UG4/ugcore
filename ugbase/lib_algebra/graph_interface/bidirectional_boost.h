/*
 * Copyright (c) 2022:  G-CSC, Goethe University Frankfurt
 * Author: Felix Salfelder, 2022
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#ifndef UG4_GRAPH_INTERFACE_BIDIR_BOOST_H
#define UG4_GRAPH_INTERFACE_BIDIR_BOOST_H

#include "bidirectional.h"

namespace boost{

struct BS_traversal_tag
    : adjacency_graph_tag, bidirectional_graph_tag, vertex_list_graph_tag {};

template <class T>
struct graph_traits<ug::BidirectionalMatrix<T>>{
	typedef typename T::value_type value_type;
	typedef int vertex_descriptor;
	typedef SM_edge<value_type> edge_descriptor;
	typedef bidirectional_tag directed_category;
	typedef BS_traversal_tag traversal_category;
	typedef disallow_parallel_edge_tag edge_parallel_category;
	typedef counting_iterator<size_t> vertex_iterator;
	typedef SM_out_edge_iterator<value_type, true> out_edge_iterator;
	typedef SM_out_edge_iterator<value_type, false> in_edge_iterator;
	typedef SM_adjacency_iterator<value_type> adjacency_iterator;
	typedef int degree_size_type;
	typedef int vertices_size_type;
};

template<class T>
std::pair<counting_iterator<size_t>, counting_iterator<size_t> > vertices(
      ug::BidirectionalMatrix<T> const& M)
{
	counting_iterator<size_t> b(0);
	counting_iterator<size_t> e(M.num_rows());

	return std::make_pair(b,e);
}

template<class T>
int num_vertices(ug::BidirectionalMatrix<T> const& M)
{
	return M.num_rows();
}

template<class T>
int out_degree(int v, ug::BidirectionalMatrix<T> const& M)
{
	return M.out_degree(v);
}

template<class T>
int in_degree(int v, ug::BidirectionalMatrix<T> const& M)
{
	return M.in_degree(v);
}

template<class T>
int degree(int v, ug::BidirectionalMatrix<T> const& M)
{ untested();
	return 2*M.degree(v);
}

template<class T>
size_t source(SM_edge<typename T::value_type> const& e, ug::BidirectionalMatrix<T> const&)
{
	return e.source();
}

template<class T>
size_t target(SM_edge<typename T::value_type> const& e, ug::BidirectionalMatrix<T> const& m)
{
	return e.target();
}

template<class T>
std::pair<typename graph_traits<T>::adjacency_iterator,
          typename graph_traits<T>::adjacency_iterator>
adjacent_vertices(size_t v, ug::BidirectionalMatrix<T> const& M)
{
	typedef typename graph_traits<T>::adjacency_iterator a;

	typename T::const_row_iterator b = M.begin_row(v);
	typename T::const_row_iterator e = M.end_row(v);

	return std::make_pair(a(&b, &e), a(&e, &e));
}

template<class T>
std::pair<typename graph_traits<T>::adjacency_iterator,
          typename graph_traits<T>::adjacency_iterator>
coadjacent_vertices(size_t v, ug::BidirectionalMatrix<T> const& M)
{
	typedef typename T::value_type value_type;
	typedef typename graph_traits<ug::SparseMatrix<value_type>>::adjacency_iterator a;

	typename T::const_row_iterator b = M.begin_col(v);
	typename T::const_row_iterator e = M.end_col(v);

	return std::make_pair(a(&b, &e), a(&e, &e));
}


template<class T>
inline std::pair<SM_out_edge_iterator<typename T::value_type>,
                 SM_out_edge_iterator<typename T::value_type>>
					out_edges(size_t v, ug::BidirectionalMatrix<T> const& g)
{
	typedef typename T::value_type value_type;
	typedef SM_out_edge_iterator<value_type> Iter;
   auto a = adjacent_vertices(v, g);
	return std::make_pair(Iter(a.first, v), Iter(a.second, v));
}

template<class T>
inline std::pair<SM_out_edge_iterator<typename T::value_type, false>,
                 SM_out_edge_iterator<typename T::value_type, false>>
					in_edges(size_t v, ug::BidirectionalMatrix<T> const& g)
{
	typedef typename T::value_type value_type;
	typedef SM_out_edge_iterator<value_type, false> Iter;
   auto a = coadjacent_vertices(v, g);
	return std::make_pair(Iter(a.first, v), Iter(a.second, v));
}

template<class T>
inline SM_edge_weight_map<typename T::value_type, ug::BidirectionalMatrix<T>>
get(edge_weight_t, ug::BidirectionalMatrix<T> const & g) {
	typedef typename T::value_type value_type;
	return SM_edge_weight_map<value_type, ug::BidirectionalMatrix<T>>(g);
}

template<class T>
struct property_map<ug::BidirectionalMatrix<ug::SparseMatrix<T>>, vertex_index_t>{
	typedef sparse_matrix_index_map<T> type;
	typedef type const_type;
};

template<class T>
inline typename property_map<ug::BidirectionalMatrix<ug::SparseMatrix<T>>, vertex_index_t>::const_type
get(vertex_index_t, ug::BidirectionalMatrix<T> const& m)
{
	return sparse_matrix_index_map<typename T::value_type>(m);
}

} // boost

#endif // guard
