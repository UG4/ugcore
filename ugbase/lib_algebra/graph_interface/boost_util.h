/*
 * Copyright (c) 2022:  G-CSC, Goethe University Frankfurt
 * Author: Felix Salfelder, 2022
 * 
 * This file is part of UG4.
 * 
 * UG4 is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 3 (as published by the
 * Free Software Foundation) with the following additional attribution
 * requirements (according to LGPL/GPL v3 §7):
 * 
 * (1) The following notice must be displayed in the Appropriate Legal Notices
 * of covered and combined works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (2) The following notice must be displayed at a prominent place in the
 * terminal output of covered works: "Based on UG4 (www.ug4.org/license)".
 * 
 * (3) The following bibliography is recommended for citation and must be
 * preserved in all covered files:
 * "Reiter, S., Vogel, A., Heppner, I., Rupp, M., and Wittum, G. A massively
 *   parallel geometric multigrid solver on hierarchically distributed grids.
 *   Computing and visualization in science 16, 4 (2013), 151-164"
 * "Vogel, A., Reiter, S., Rupp, M., Nägel, A., and Wittum, G. UG4 -- a novel
 *   flexible software system for simulating pde based models on high performance
 *   computers. Computing and visualization in science 16, 4 (2013), 165-179"
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 */

#ifndef UG_GRAPH_INTERFACE_BOOST_UTIL_H
#define UG_GRAPH_INTERFACE_BOOST_UTIL_H

#include <utility>
#include <boost/iterator/filter_iterator.hpp>

namespace ug{
namespace util{

namespace{
template<class G>
class noloop{
public:
	explicit noloop(const G& g) : _g(g) {}
	template<class E>
	bool operator()(E const&e) const{
		return boost::source(e, _g) != boost::target(e, _g);
	}
private:
	G const& _g;
};
}

template<class T, class G>
std::pair<boost::filter_iterator<noloop<G>, T>,
          boost::filter_iterator<noloop<G>, T>> omit_loops(std::pair<T, T> const& p, G const& g)
{
	noloop<G> P(g);
	typedef boost::filter_iterator<noloop<G>, T> f;
	return std::make_pair(f(P, p.first, p.second), f(P, p.second, p.second));
}

} // util
} // ug

#endif // guard
