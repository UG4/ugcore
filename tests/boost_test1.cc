
#include "lib_algebra/graph_interface/sparsematrix_boost.h"
#include "lib_algebra/cpu_algebra/sparsematrix_impl.h"
#include "lib_algebra/graph_interface/bidirectional_boost.h"
#include "lib_algebra/graph_interface/boost_util.h"

#include "common/log.cpp" // ?
#include "common/debug_id.cpp" // ?
#include "common/assert.cpp" // ?
#include "common/util/crc32.cpp" // ?
#include "common/util/ostream_buffer_splitter.cpp" // ?
#include "common/util/string_util.cpp" // ?

#include "boost/graph/graph_utility.hpp"

// bidirectional tests

static const int N=7;

void test1()
{
	typedef ug::SparseMatrix<double> T;
	typedef ug::BidirectionalMatrix<T> BM;
	ug::SparseMatrix<double> M;

	typedef typename boost::graph_traits<BM>::in_edge_iterator in_edge_iterator;
	typedef typename boost::graph_traits<BM>::out_edge_iterator out_edge_iterator;

	M.resize_and_clear(N, N+2);

	M(0, 0) = 2.;
	M(1, N-1) = -1.;
	for(unsigned i=1; i<N; ++i){
		M(i, i) = 2.;
		M(i, i-1) = 1.;
		M(i-1, i) = 1.;
	}
	M(N-1, N+1) = -1.;

	BM b_(&M);
	BM b;
	b = b_;

	{
		auto e = boost::vertices(b);
		for(; e.first!=e.second; ++e.first){
			auto v = *e.first;
			std::cout << v << " outdeg: " << boost::out_degree(v, b) << " indeg: " << boost::in_degree(v, b) << "\n";
		}
	}

	auto wtmap = boost::get(boost::edge_weight, b);

	for(int w=0; w<N+2; ++w){
		std::cout << "=== out edges " << w << " ===\n";
		std::pair<out_edge_iterator, out_edge_iterator> e = boost::out_edges(w, b);
		for(; e.first!=e.second; ++e.first){
			boost::graph_traits<BM>::edge_descriptor const& edg = *e.first;
			int s = boost::source(edg, b);
			assert(w == s);
			std::cout << s << "->" << boost::target(edg, b)
			          << " -- " << boost::get(wtmap, edg) << "\n";
		}
	}

	for(int w=0; w<N+2; ++w){
		std::cout << "=== in edges === " << w << "\n";
		std::pair<in_edge_iterator, in_edge_iterator> e = boost::in_edges(w, b);
		for(; e.first!=e.second; ++e.first){
			boost::graph_traits<BM>::edge_descriptor const& edg = *e.first;
			int t = boost::target(edg, b);
			std::cout << boost::source(edg, b) << "->" << t
			          << " -- " << boost::get(wtmap, edg) << "\n";
			assert(w == t);
		}
	}

	auto n = boost::num_vertices(b);

	std::cout << "pg " << n << "\n";
	boost::print_graph(b);
}

void test0()
{
	typedef ug::SparseMatrix<double> T;
	typedef ug::BidirectionalMatrix<T> BM;
	ug::SparseMatrix<double> M;

	typedef typename boost::graph_traits<BM>::in_edge_iterator in_edge_iterator;
	typedef typename boost::graph_traits<BM>::out_edge_iterator out_edge_iterator;

	M.resize_and_clear(N, N);

	M(0, 0) = 2.;
	M(0, 6) = -1.;
	M(5, 1) = 0.;
	for(unsigned i=1; i<N; ++i){
		M(i, i) = 2.;
		M(i, i-1) = 1.;
		M(i-1, i) = 1.;
	}

	BM b_(&M);
	BM b(b_);

	{
		auto e = boost::vertices(b);
		for(; e.first!=e.second; ++e.first){
			auto v = *e.first;
			std::cout << v << " " << boost::out_degree(v, b) << " " << boost::in_degree(v, b) << "\n";
		}
	}

	auto wtmap = boost::get(boost::edge_weight, b);

	{
		for(int k=0; k<N; ++k){
			std::cout << "=== out edges " << k << " === "
			          << boost::out_degree(k, b) << "\n";
			std::pair<out_edge_iterator, out_edge_iterator> e = boost::out_edges(k, b);
			// auto e = ug::util::omit_loops(e_, b);
			for(; e.first!=e.second; ++e.first){
				boost::graph_traits<BM>::edge_descriptor const& edg = *e.first;
				std::cout << boost::source(edg, b) << ":" << boost::target(edg, b)
							 << " -- " << boost::get(wtmap, edg) << "\n";
			}
		}
	}

	if(0){
		std::cout << "=== out edges ===\n";
		std::pair<out_edge_iterator, out_edge_iterator> e_ = boost::out_edges(5, b);
		auto e = ug::util::omit_loops(e_, b);
		for(; e.first!=e.second; ++e.first){
			boost::graph_traits<BM>::edge_descriptor const& edg = *e.first;
			auto wt = boost::get(wtmap, edg);
			std::cout << boost::source(edg, b) << ":" << boost::target(edg, b)
			          << " -- " << wt << "\n";
			assert(wt);
		}
	}

	{
		std::cout << "=== in edges ===\n";
		std::pair<in_edge_iterator, in_edge_iterator> e = boost::in_edges(1, b);
		for(; e.first!=e.second; ++e.first){
			boost::graph_traits<BM>::edge_descriptor const& edg = *e.first;
			std::cout << boost::source(edg, b) << ":" << boost::target(edg, b)
			          << " -- " << boost::get(wtmap, edg) << "\n";
		}
	}
}

int main()
{
	std::cout << "== test0\n";
	test0();
	std::cout << "== test1\n";
	test1();
}

typedef ug::DenseMatrix<ug::VariableArray2<double> > T;
BOOST_CONCEPT_ASSERT(( boost::AdjacencyGraphConcept<ug::SparseMatrix<T>> ));
BOOST_CONCEPT_ASSERT(( boost::IncidenceGraphConcept<ug::SparseMatrix<T>> ));
BOOST_CONCEPT_ASSERT(( boost::VertexListGraphConcept<ug::SparseMatrix<T>> ));

BOOST_CONCEPT_ASSERT(( boost::BidirectionalGraphConcept<ug::BidirectionalMatrix<ug::SparseMatrix<double>>> ));
BOOST_CONCEPT_ASSERT(( boost::AdjacencyGraphConcept<ug::BidirectionalMatrix<ug::SparseMatrix<double>>> ));
BOOST_CONCEPT_ASSERT(( boost::VertexListGraphConcept<ug::BidirectionalMatrix<ug::SparseMatrix<double>>> ));

