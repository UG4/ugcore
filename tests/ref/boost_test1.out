== test0
0 3 2
1 3 3
2 3 3
3 3 3
4 3 3
5 3 3
6 2 3
=== out edges 0 === 3
0:0 -- 2
0:1 -- 1
0:6 -- -1
=== out edges 1 === 3
1:0 -- 1
1:1 -- 2
1:2 -- 1
=== out edges 2 === 3
2:1 -- 1
2:2 -- 2
2:3 -- 1
=== out edges 3 === 3
3:2 -- 1
3:3 -- 2
3:4 -- 1
=== out edges 4 === 3
4:3 -- 1
4:4 -- 2
4:5 -- 1
=== out edges 5 === 3
5:4 -- 1
5:5 -- 2
5:6 -- 1
=== out edges 6 === 2
6:5 -- 1
6:6 -- 2
=== in edges ===
0:1 -- 1
1:1 -- 2
2:1 -- 1
== test1
0 outdeg: 2 indeg: 2
1 outdeg: 4 indeg: 3
2 outdeg: 3 indeg: 3
3 outdeg: 3 indeg: 3
4 outdeg: 3 indeg: 3
5 outdeg: 3 indeg: 3
6 outdeg: 3 indeg: 3
7 outdeg: 0 indeg: 0
8 outdeg: 0 indeg: 1
=== out edges 0 ===
0->0 -- 2
0->1 -- 1
=== out edges 1 ===
1->0 -- 1
1->1 -- 2
1->2 -- 1
1->6 -- -1
=== out edges 2 ===
2->1 -- 1
2->2 -- 2
2->3 -- 1
=== out edges 3 ===
3->2 -- 1
3->3 -- 2
3->4 -- 1
=== out edges 4 ===
4->3 -- 1
4->4 -- 2
4->5 -- 1
=== out edges 5 ===
5->4 -- 1
5->5 -- 2
5->6 -- 1
=== out edges 6 ===
6->5 -- 1
6->6 -- 2
6->8 -- -1
=== out edges 7 ===
=== out edges 8 ===
=== in edges === 0
0->0 -- 2
1->0 -- 1
=== in edges === 1
0->1 -- 1
1->1 -- 2
2->1 -- 1
=== in edges === 2
1->2 -- 1
2->2 -- 2
3->2 -- 1
=== in edges === 3
2->3 -- 1
3->3 -- 2
4->3 -- 1
=== in edges === 4
3->4 -- 1
4->4 -- 2
5->4 -- 1
=== in edges === 5
4->5 -- 1
5->5 -- 2
6->5 -- 1
=== in edges === 6
1->6 -- -1
5->6 -- 1
6->6 -- 2
=== in edges === 7
=== in edges === 8
6->8 -- -1
pg 9
0 --> 0 1 
1 --> 0 1 2 6 
2 --> 1 2 3 
3 --> 2 3 4 
4 --> 3 4 5 
5 --> 4 5 6 
6 --> 5 6 8 
7 --> 
8 --> 
